sno/working_copy/sqlserver.py

```diff
@@ -0,0 +1,553 @@
+        new_type = new_col_dict["dataType"]
+
+        # Some types have to be approximated as other types in SQL Server.
+        if sqlserver_adapter.APPROXIMATED_TYPES.get(old_type) == new_type:
+            new_col_dict["dataType"] = new_type = old_type
+
+        # Geometry type loses its extra type info when roundtripped through SQL Server.
+        if new_type == "geometry":
+            new_col_dict["geometryType"] = old_col_dict.get("geometryType")
+            new_col_dict["geometryCRS"] = old_col_dict.get("geometryCRS")
+
+        return new_type == old_type
+
+    def _remove_hidden_meta_diffs(self, dataset, ds_meta_items, wc_meta_items):
+        super()._remove_hidden_meta_diffs(dataset, ds_meta_items, wc_meta_items)
+
+        # Nowhere to put these in postgis WC
+        for key in self._UNSUPPORTED_META_ITEMS:
+            if key in ds_meta_items:
+                del ds_meta_items[key]
+
+        # Diffing CRS is not yet supported.
+        for key in list(ds_meta_items.keys()):
+            if key.startswith("crs/"):
+                del ds_meta_items[key]
+
+    def _is_meta_update_supported(self, dataset_version, meta_diff):
+        """
+        Returns True if the given meta-diff is supported *without* dropping and rewriting the table.
+        (Any meta change is supported if we drop and rewrite the table, but of course it is less efficient).
+        meta_diff - DeltaDiff object containing the meta changes.
+        """
+        # For now, just always drop and rewrite.
+        return not meta_diff
+
+
+class InstanceFunction(Function):
+    """
+    An instance function that compiles like this when applied to an element:
+    >>> element.function()
+    Unlike a normal sqlalchemy function which would compile as follows:
+    >>> function(element)
+    """
+
+    pass
+
+
+@compiles(InstanceFunction)
+def compile_instance_function(element, compiler, **kw):
+    return "(%s).%s()" % (element.clauses, element.name)
+
+
+class GeographyType(UserDefinedType):
+    """UserDefinedType so that V2 geometry is adapted to MS binary format."""
+
+    def __init__(self, crs_id):
+        self.crs_id = crs_id
+
+    def bind_processor(self, dialect):
+        # 1. Writing - Python layer - convert sno geometry to WKB
+        return lambda geom: geom.to_wkb()
+
+    def bind_expression(self, bindvalue):
+        # 2. Writing - SQL layer - wrap in call to STGeomFromWKB to convert WKB to MS binary.
+        return Function(
+            quoted_name("geography::STGeomFromWKB", False),
+            bindvalue,
+            self.crs_id,
+            type_=self,
+        )
+
+    def column_expression(self, col):
+        # 3. Reading - SQL layer - append with call to .STAsBinary() to convert MS binary to WKB.
+        return InstanceFunction("STAsBinary", col, type_=self)
+
+    def result_processor(self, dialect, coltype):
+        # 4. Reading - Python layer - convert WKB to sno geometry.
+        return lambda wkb: Geometry.from_wkb(wkb)
```

Comment:
This needs to be lazy too.

Reply:
Does MSSQL have a concept of a per-geometry CRS/SRID like PostGIS does? If so, should `Geometry` be initialised with the feature or WC SRID?

Reply:
SQL server does have a concept of per-geometry CRS - in fact that is the only concept it has, unlike PostGIS, there is no column- or table-level way to specify the CRS.

In Sno on the other hand we *only* support column-level CRS (roughly equivalent to table-level, since there is normally only one geometry column). We specifically zero-out the CRS's in the individual GPKG geometries on commit so that you can't have different ones per geometry, and we treat them all as having the column-level one.

So, when reading geometries from the working copy, we always remove any embedded CRS ID - regardless of working copy type. Geometries can then be diffed against the stored Sno geometries, independently of CRS, and the CRS diff (if there is one) will show up only in the meta diff - since, Sno the only CRS change that Sno lets you diff and commit is the one where you change the entire column.

This is a piece of missing functionality in SQL Server right now - there is no way to change the CRS of a column except using `sno meta set`, since there is no other easy way for the user to signal to us that they want to change a columns CRS. Perhaps we should do a heuristic where we (for instance) sample a few geometries and see if they all have the same CRS - if so, that is probably the user-selected column CRS, and we can generate a diff.

Apologies, this is all undocumented since I haven't written SQL_SERVER.md or whatever it will be called

Reply:
> unlike PostGIS, there is no column- or table-level way to specify the CRS.

PostGIS supports both... per-geometry CRS is normal, with a table-level constraint to check it. But without the constraint it'll let anything be stored.

> Perhaps we should do a heuristic where we (for instance) sample a few geometries and see if they all have the same CRS - if so, that is probably the user-selected column CRS, and we can generate a diff.

kinda-sometimes-working magic is worse IME than no magic. It's ok we only support homogenous CRS in a single dataset — that is by far the most normal scenario. And as long as we have some steps to follow to change the CRS then that's good.

Reply:
Marking this as resolved